sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadTableSource.java

```diff
@@ -0,0 +1,216 @@
+ */
+class BigQueryParallelReadTableSource<T> extends BoundedSource<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryParallelReadTableSource.class);
+
+  private BigQueryServices bqServices;
+  private Coder<T> coder;
+  private SerializableFunction<SchemaAndRowProto, T> parseFn;
+  private final AtomicReference<Long> tableSizeBytes;
+  private Session session;
+  private ValueProvider<String> jsonTable;
+  private ReadLocation initialReadLocation;
+  private ReadSessionOptions readOptions;
+  private transient List<BoundedSource<T>> cachedSplitResult;
+
+  private BigQueryParallelReadTableSource(ValueProvider<String> jsonTable,
+                                Session session,
+                                ReadLocation readLocation,
+                                SerializableFunction<SchemaAndRowProto, T> parseFn,
+                                Coder<T> coder,
+                                BigQueryServices bqServices,
+                                ReadSessionOptions readOptions) {
+    this.bqServices = checkNotNull(bqServices, "bqServices");
+    this.coder = checkNotNull(coder, "coder");
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.jsonTable = checkNotNull(jsonTable, "jsonTable");
+    this.session = session;
+    this.tableSizeBytes = new AtomicReference<>();
+    this.initialReadLocation = readLocation;
+    this.readOptions = readOptions;
+  }
+
+  /**
+   * This method creates a {@link BigQueryParallelReadTableSource} with no initial session or read
+   * location.
+   */
+  public static <T> BigQueryParallelReadTableSource<T> create(
+      ValueProvider<TableReference> table,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      BigQueryServices bqServices,
+      ReadSessionOptions readOptions) {
+    return new BigQueryParallelReadTableSource<>(
+        NestedValueProvider.of(checkNotNull(table, "table"), new TableRefToJson()),
+        null,
+        // First TableSource shouldn't be used to CreateReader.
+        ReadLocation.getDefaultInstance(),
+        parseFn,
+        coder,
+        checkNotNull(bqServices),
+        readOptions);
+  }
+
+  /*
+   * Gets the total size of the table.
+   */
+  @Override
+  public synchronized long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+    if (tableSizeBytes.get() == null) {
+      TableReference table =
+          BigQueryIO.JSON_FACTORY.fromString(jsonTable.get(), TableReference.class);
+
+      Table tableRef = bqServices.getDatasetService(options.as(BigQueryOptions.class))
+          .getTable(table);
+      Long numBytes = tableRef.getNumBytes();
+      if (tableRef.getStreamingBuffer() != null) {
+        numBytes += tableRef.getStreamingBuffer().getEstimatedBytes().longValue();
+      }
+
+      tableSizeBytes.compareAndSet(null, numBytes);
+    }
+    return tableSizeBytes.get();
+  }
+
+  private static com.google.cloud.bigquery.v3.TableReferenceProto.TableReference
```

Comment:
Can these imports be shortened?

Reply:
We're mixing the v2 and v3 APIs here, and both APIs have their own TableReference types. I suppose we could shorten this to TableReferenceProto.TableReference.